isovar/genetic_code.py

```diff
@@ -0,0 +1,165 @@
+
+class GeneticCode(object):
+    def __init__(self, name, start_codons, stop_codons, codon_table):
+        self.name = name
+        self.start_codons = set(start_codons)
+        self.stop_codons = set(stop_codons)
+        self.codon_table = dict(codon_table)
+        self._check_codons()
+
+    def _check_codons(self):
+        """
+        If codon table is missing stop codons, then add them.
+        """
+        for stop_codon in self.stop_codons:
+            if stop_codon in self.codon_table:
+                if self.codon_table[stop_codon] != "*":
+                    raise ValueError(
+                        "Stop codon '%s' must translate to '*'" % (
+                            stop_codon,))
+            else:
+                self.codon_table[stop_codon] = "*"
+        for start_codon in self.start_codons:
+            if start_codon not in self.codon_table:
+                raise ValueError(
+                    "Start codon '%s' missing from codon table" % (
+                        start_codon,))
+        for codon, amino_acid in self.codon_table.items():
+            if amino_acid == "*" and codon not in self.stop_codons:
+                raise ValueError(
+                    "Non-stop codon '%s' can't translate to '*'" % (
+                        codon,))
+        if len(self.codon_table) != 64:
+            raise ValueError(
+                "Expected 64 codons but found %d in codon table" % (
+                    len(self.codon_table,)))
+
+    def translate(self, cdna_sequence, first_codon_is_start=False):
+        """
+        Given a cDNA sequence which is aligned to a reading frame, returns
+        the translated protein sequence and a boolean flag indicating whether
+        the translated sequence ended on a stop codon (or just ran out of codons).
+
+        Parameters
+        ----------
+        cdna_sequence : str
+            cDNA sequence which is expected to start and end on complete codons.
+
+        first_codon_is_start : bool
+            Is the first codon of the sequence a start codon?
+        """
+        if not isinstance(cdna_sequence, str):
+            cdna_sequence_from_first_codon = str(cdna_sequence)
+        n = len(cdna_sequence_from_first_codon)
+        if len(cdna_sequence_from_first_codon) < 3:
+            return ''
+
+        codon_table = self.codon_table
+        if first_codon_is_start and cdna_sequence[:3] in self.start_codons:
+            amino_acid_list = ['M']
+            start_index = 3
+        else:
+            start_index = 0
+            amino_acid_list = []
+
+        # trim to multiple of 3 length
+        end_idx = 3 * (n // 3)
```

Comment:
I don't think it's an error really. If an mRNA is missing its stop codon then the ribosome eventually falls off -- regardless of the sequence length. And if there's a stop codon before the last incomplete codon then the ribosome falls off there.